File: extension/core/src/main/java/io/holunda/camunda/bpm/data/adapter/list/ListReadAdapterLockedExternalTask.java
package io.holunda.camunda.bpm.data.adapter.list;

import io.holunda.camunda.bpm.data.adapter.ReadAdapter;
import io.holunda.camunda.bpm.data.adapter.WrongVariableTypeException;
import org.camunda.bpm.engine.externaltask.LockedExternalTask;

import java.util.Collections;
import java.util.List;
import java.util.Optional;

/**
 * Read adapter for external task.
 *
 * @param <T> type of value.
 */
@SuppressWarnings("unchecked")
public class ListReadAdapterLockedExternalTask<T> implements ReadAdapter<List<T>> {

  private final LockedExternalTask lockedExternalTask;
  private final String variableName;
  private final Class<T> memberClazz;

  public ListReadAdapterLockedExternalTask(LockedExternalTask lockedExternalTask, String variableName, Class<T> memberClazz) {
    this.lockedExternalTask = lockedExternalTask;
    this.variableName = variableName;
    this.memberClazz = memberClazz;
  }

  @Override
  public List<T> get() {
    return null;
  }

  @Override
  @SuppressWarnings("java:S3655")
  public Optional<List<T>> getOptional() {
    return Optional.ofNullable(getOrNull(getValue()));
  }

  @Override
  public List<T> getLocal() {
    throw new UnsupportedOperationException("Can't get a local variable on an external task");
  }

  @Override
  public Optional<List<T>> getLocalOptional() {
    throw new UnsupportedOperationException("Can't get a local variable on an external task");
  }

  @Override
  public List<T> getOrDefault(List<T> defaultValue) {
    return getOptional().orElse(defaultValue);
  }

  @Override
  public List<T> getLocalOrDefault(List<T> defaultValue) {
    throw new UnsupportedOperationException("Can't get a local variable on an external task");
  }

  @Override
  public List<T> getOrNull() {
    return getOrNull(getValue());
  }

  @Override
  public List<T> getLocalOrNull() {
    throw new UnsupportedOperationException("Can't get a local variable on an external task");
  }

  @SuppressWarnings("Duplicates")
  private List<T> getOrNull(T value) {
    if (value == null) {
      return null;
    }

    if (List.class.isAssignableFrom(value.getClass())) {
      List<?> valueAsList = (List<?>) value;
      if (valueAsList.isEmpty()) {
        return Collections.emptyList();
      } else {
        if (memberClazz.isAssignableFrom(valueAsList.iterator().next().getClass())) {
          return (List<T>) valueAsList;
        } else {
          throw new WrongVariableTypeException("Error reading " + variableName + ": Wrong list type detected, expected " + memberClazz.getName() + ", but was not found in " + valueAsList);
        }
      }
    }

    throw new WrongVariableTypeException("Error reading " + variableName + ": Couldn't read value of type List from " + value);
  }

  private T getValue() {
    return (T) Optional.ofNullable(lockedExternalTask.getVariables()).map(it -> it.get(variableName)).get();
  }
}

File: extension/core/src/main/kotlin/io/holunda/camunda/bpm/data/adapter/list/ListReadAdapterLockedExternalTask.kt
package io.holunda.camunda.bpm.data.adapter.list

import io.holunda.camunda.bpm.data.adapter.ReadAdapter
import io.holunda.camunda.bpm.data.adapter.WrongVariableTypeException
import org.camunda.bpm.engine.externaltask.LockedExternalTask
import org.camunda.bpm.engine.variable.Variables
import java.util.*

/**
 * Read adapter for external task.
 *
 * @param [T] type of value.
 */
class ListReadAdapterLockedExternalTask<T>(
    private val lockedExternalTask: LockedExternalTask,
    private val variableName: String,
    private val memberClazz: Class<T>
) : ReadAdapter<List<T>> {

    override fun get(): List<T> {
        return getOptional().get()
    }

    override fun getOptional(): Optional<List<T>> {
        return Optional.ofNullable(getOrNull(value))
    }

    override fun getLocal(): List<T> {
        throw UnsupportedOperationException("Can't get a local variable on an external task")
    }

    override fun getLocalOptional(): Optional<List<T>> {
        throw UnsupportedOperationException("Can't get a local variable on an external task")
    }

    override fun getOrDefault(defaultValue: List<T>): List<T> {
        return getOptional().orElse(defaultValue)
    }

    override fun getLocalOrDefault(defaultValue: List<T>): List<T> {
        throw UnsupportedOperationException("Can't get a local variable on an external task")
    }

    override fun getOrNull(): List<T>? {
        return getOrNull(value)
    }

    override fun getLocalOrNull(): List<T> {
        throw UnsupportedOperationException("Can't get a local variable on an external task")
    }

    private fun getOrNull(value: T?): List<T>? {
        if (value == null) {
            return null
        }
        if (MutableList::class.java.isAssignableFrom(value.javaClass)) {
            val valueAsList = value as List<*>
            return if (valueAsList.isEmpty()) {
                emptyList()
            } else {
                if (memberClazz.isAssignableFrom(valueAsList.iterator().next()!!.javaClass)) {
                    valueAsList as List<T>
                } else {
                    throw WrongVariableTypeException("Error reading " + variableName + ": Wrong list type detected, expected " + memberClazz.name + ", but was not found in " + valueAsList)
                }
            }
        }
        throw WrongVariableTypeException("Error reading $variableName: Couldn't read value of type List from $value")
    }

    private val value: T?
        get() = Optional.ofNullable(lockedExternalTask.variables)
            .orElse(Variables.createVariables())[variableName] as T?
}
